### PR TITLE
Add BE plugin framework (#2348)

### PR DESCRIPTION
**Pull request by Seaven**
support BE plugin framework, include:

* update Plugin Manager, support Plugin find method

* support Builtin-Plugin register method

* plugin install/uninstall process
	* PluginLoader:
		* dynamic install and check Plugin .so file
		* dynamic uninstall and check Plugin status
		
	* PluginZip:
		* support plugin remote/local .zip file download and extract

and Todo:

* We should support a PluginContext to transmit necessary system variable when the plugin's init/close method invoke
* Add the entry which is BE dynamic Plugin install/uninstall process, include:

	* The FE send install/uninstall Plugin statement (RPC way)
	* The FE meta update request with Plugin list information
	* The FE operation request(update/query) with Plugin (maybe don't need)

* Add the plugin status upload way
* Load already install Plugin when BE start

### CODE REVIEW

File: be/src/plugin/plugin.h [hunk 1]
```diff
@@ -15,11 +15,17 @@
...
 
+#ifndef DORIS_BE_PLUGIN_PLUGIN_H
+#define DORIS_BE_PLUGIN_PLUGIN_H
+
 namespace doris {
 
-#define PLUGIN_NOT_DYNAMIC_INSTALL 1UL
+#define PLUGIN_TYPE_IMPORT 1
+#define PLUGIN_TYPE_STORAGE 2
+#define PLUGIN_TYPE_MAX 3
```
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/plugin/plugin.h]: Temporary, we can add later

File: gensrc/thrift/AgentService.thrift [hunk 2]
```diff
@@ -279,6 +294,8 @@ struct TAgentTaskRequest {
     24: optional TAlterTabletReqV2 alter_tablet_req_v2
     25: optional i64 recv_time // time the task is inserted to queue
     26: optional TUpdateTabletMetaInfoReq update_tablet_meta_info_req
+    27: optional TInstallPluginReq install_plugin_req
```
Comment (morningman) [inline gensrc/thrift/AgentService.thrift]: You can currently remove these 2 fields, add them when implement them.

File: gensrc/thrift/MasterService.thrift [hunk 3]
```diff
@@ -84,6 +89,8 @@ struct TReportRequest {
     // the max compaction score of all tablets on a backend,
     // this field should be set along with tablet report
     8: optional i64 tablet_max_compaction_score
+    // plugin list
+    9: optional list<TPluginInfo> plugins
```
Comment (morningman) [inline gensrc/thrift/MasterService.thrift]: Remove this field, add it later

File: be/src/plugin/plugin_mgr.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    
+    PluginMgr() {}
+    
+    ~PluginMgr() {}
+    
+    Status install_plugin(const TPluginMetaInfo& info);
+    
+    Status uninstall_plugin(const TPluginMetaInfo& info);
+
+    Status register_builtin_plugin(const std::string& name, int type, Plugin* plugin);
```
Comment (morningman) [inline be/src/plugin/plugin_mgr.h]: Can the parameter be : `const Plugin* plugin`?

File: be/src/plugin/plugin_loader.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    if (plugin_size != sizeof(Plugin)) {
+        return Status::InternalError("plugin struct error");
+    }
+
+    // check Plugin declaration
+    RETURN_IF_ERROR(dynamic_lookup(_plugin_handler, (_name + PLUGIN_STRUCT_SYMBOL).c_str(), &symbol));
+
+    Plugin* end_plugin = (Plugin*) ((char*) symbol + plugin_size);
+
+    if (end_plugin->handler != nullptr || end_plugin->init != nullptr || end_plugin->close != nullptr) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/plugin/plugin_loader.cpp]: For check plugin struct

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**